File: CarrelApp/Core/Network/ConvexClient.swift
import Foundation

actor ConvexClient {
    private let baseURL: URL
    private let authManager: AuthManager
    private let session: URLSession
    private let decoder: JSONDecoder

    init(baseURL: URL, authManager: AuthManager) {
        self.baseURL = baseURL
        self.authManager = authManager

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        self.session = URLSession(configuration: config)

        self.decoder = JSONDecoder()
    }

    // MARK: - Papers

    func papers() async throws -> [Paper] {
        let data: [Paper] = try await request(.get, "/api/mobile/papers")
        return data
    }

    func paper(id: String) async throws -> Paper {
        try await request(.get, "/api/mobile/paper", query: ["id": id])
    }

    func buildPaper(id: String, force: Bool = false) async throws {
        let _: EmptyResponse = try await request(
            .post,
            "/api/mobile/paper/build",
            body: ["paperId": id, "force": force]
        )
    }

    func deletePaper(id: String) async throws {
        let _: EmptyResponse = try await request(
            .delete,
            "/api/mobile/paper",
            body: ["paperId": id]
        )
    }

    func updatePaper(id: String, title: String?, authors: String?) async throws {
        var body: [String: Any] = ["paperId": id]
        if let title = title {
            body["title"] = title
        }
        if let authors = authors {
            body["authors"] = authors
        }
        let _: EmptyResponse = try await request(.patch, "/api/mobile/paper", body: body)
    }

    func togglePaperPublic(id: String) async throws -> TogglePublicResponse {
        try await request(
            .post,
            "/api/mobile/paper/toggle-public",
            body: ["paperId": id]
        )
    }

    // MARK: - User

    func user() async throws -> User {
        try await request(.get, "/api/mobile/user")
    }

    // MARK: - Auth

    func refreshToken(_ refreshToken: String) async throws -> RefreshTokenResponse {
        try await requestWithoutAuth(
            .post,
            "/api/mobile/refresh",
            body: ["refreshToken": refreshToken]
        )
    }

    func revokeToken(_ refreshToken: String) async throws {
        let _: EmptyResponse = try await requestWithoutAuth(
            .post,
            "/api/mobile/revoke",
            body: ["refreshToken": refreshToken]
        )
    }

    // MARK: - Private

    private func request<T: Decodable>(
        _ method: HTTPMethod,
        _ path: String,
        query: [String: String]? = nil,
        body: [String: Any]? = nil
    ) async throws -> T {
        let token = try await authManager.getValidToken()
        return try await performRequest(method, path, query: query, body: body, token: token)
    }

    private func requestWithoutAuth<T: Decodable>(
        _ method: HTTPMethod,
        _ path: String,
        query: [String: String]? = nil,
        body: [String: Any]? = nil
    ) async throws -> T {
        try await performRequest(method, path, query: query, body: body, token: nil)
    }

    private func performRequest<T: Decodable>(
        _ method: HTTPMethod,
        _ path: String,
        query: [String: String]?,
        body: [String: Any]?,
        token: String?
    ) async throws -> T {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: true)

        if let query = query {
            components?.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        }

        guard let url = components?.url else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method.rawValue

        if let token = token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body = body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.unknown(0, "Invalid response type")
        }

        switch httpResponse.statusCode {
        case 200...299:
            do {
                return try decoder.decode(T.self, from: data)
            } catch {
                throw APIError.decodingError(error)
            }

        case 401:
            // Try to parse error message
            if let errorResponse = try? decoder.decode(APIErrorResponse.self, from: data) {
                if errorResponse.error.contains("expired") {
                    throw APIError.tokenExpired
                }
            }
            throw APIError.unauthorized

        case 404:
            throw APIError.notFound

        case 400:
            let message = (try? decoder.decode(APIErrorResponse.self, from: data))?.error ?? "Bad request"
            throw APIError.badRequest(message)

        case 500...599:
            let message = (try? decoder.decode(APIErrorResponse.self, from: data))?.error ?? "Server error"
            throw APIError.serverError(message)

        default:
            let message = (try? decoder.decode(APIErrorResponse.self, from: data))?.error
            throw APIError.unknown(httpResponse.statusCode, message)
        }
    }
}

enum HTTPMethod: String {
    case get = "GET"
    case post = "POST"
    case patch = "PATCH"
    case delete = "DELETE"
}

struct EmptyResponse: Codable {
    let success: Bool?
}

struct TogglePublicResponse: Codable {
    let isPublic: Bool
    let shareSlug: String?
}

File: CarrelAndroid/app/src/main/java/com/carrel/app/core/network/ConvexClient.kt
package com.carrel.app.core.network

import android.util.Log
import com.carrel.app.core.auth.AuthManager
import com.carrel.app.core.network.models.*
import java.net.URLEncoder
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.android.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json

class ConvexClient(
    private val authManager: AuthManager
) {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val httpClient = HttpClient(Android) {
        install(ContentNegotiation) {
            json(json)
        }
        install(HttpTimeout) {
            requestTimeoutMillis = 30_000
            connectTimeoutMillis = 15_000
        }
    }

    // MARK: - Papers

    suspend fun papers(): ApiResult<List<Paper>> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return get("$BASE_URL/api/mobile/papers", token)
    }

    suspend fun paper(id: String): ApiResult<Paper> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return get("$BASE_URL/api/mobile/paper?id=$id", token)
    }

    suspend fun buildPaper(id: String, force: Boolean = false): ApiResult<SuccessResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return post("$BASE_URL/api/mobile/paper/build", token, mapOf("paperId" to id, "force" to force))
    }

    suspend fun deletePaper(id: String): ApiResult<SuccessResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return delete("$BASE_URL/api/mobile/paper", token, mapOf("paperId" to id))
    }

    suspend fun updatePaper(
        id: String,
        title: String? = null,
        authors: String? = null
    ): ApiResult<SuccessResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        val body = buildMap<String, Any> {
            put("paperId", id)
            title?.let { put("title", it) }
            authors?.let { put("authors", it) }
        }
        return patch("$BASE_URL/api/mobile/paper", token, body)
    }

    suspend fun togglePaperPublic(id: String): ApiResult<TogglePublicResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return post("$BASE_URL/api/mobile/paper/toggle-public", token, mapOf("paperId" to id))
    }

    // MARK: - Repositories

    suspend fun repositories(): ApiResult<List<Repository>> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return get("$BASE_URL/api/mobile/repositories", token)
    }

    suspend fun refreshRepository(id: String): ApiResult<RefreshRepositoryResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return post("$BASE_URL/api/mobile/repository/refresh", token, mapOf("repositoryId" to id))
    }

    suspend fun deleteRepository(id: String): ApiResult<SuccessResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return delete("$BASE_URL/api/mobile/repository", token, mapOf("repositoryId" to id))
    }

    suspend fun checkAllRepositories(): ApiResult<CheckAllResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return post("$BASE_URL/api/mobile/repositories/check-all", token, null)
    }

    suspend fun listRepositoryFiles(
        gitUrl: String,
        path: String?,
        branch: String?
    ): ApiResult<List<RepositoryFile>> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        val params = buildString {
            append("gitUrl=${java.net.URLEncoder.encode(gitUrl, "UTF-8")}")
            path?.let { append("&path=${java.net.URLEncoder.encode(it, "UTF-8")}") }
            branch?.let { append("&branch=${java.net.URLEncoder.encode(it, "UTF-8")}") }
        }
        return get("$BASE_URL/api/mobile/repository/files?$params", token)
    }

    suspend fun listTrackedFiles(repositoryId: String): ApiResult<List<TrackedFileInfo>> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return get("$BASE_URL/api/mobile/repository/tracked-files?repositoryId=$repositoryId", token)
    }

    suspend fun addTrackedFile(
        repositoryId: String,
        filePath: String,
        title: String,
        pdfSourceType: String,
        compiler: String?
    ): ApiResult<AddTrackedFileResponse> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        val body = buildMap<String, Any> {
            put("repositoryId", repositoryId)
            put("filePath", filePath)
            put("title", title)
            put("pdfSourceType", pdfSourceType)
            compiler?.let { put("compiler", it) }
        }
        return post("$BASE_URL/api/mobile/repository/add-tracked-file", token, body)
    }

    // MARK: - User

    suspend fun user(): ApiResult<User> {
        val token = authManager.getValidToken() ?: return ApiResult.Error(ApiException.TokenExpired)
        return get("$BASE_URL/api/mobile/user", token)
    }

    // MARK: - Auth (no auth required)

    suspend fun loginWithEmail(
        email: String,
        password: String,
        deviceId: String? = null,
        deviceName: String? = null,
        platform: String = "android"
    ): ApiResult<AuthTokens> {
        return post(
            "$BASE_URL/api/mobile/auth/email",
            null,
            mapOf(
                "email" to email,
                "password" to password,
                "deviceId" to (deviceId ?: ""),
                "deviceName" to (deviceName ?: "Android Device"),
                "platform" to platform
            )
        )
    }

    suspend fun refreshToken(refreshToken: String): ApiResult<RefreshTokenResponse> {
        return post("$BASE_URL/api/mobile/refresh", null, mapOf("refreshToken" to refreshToken))
    }

    suspend fun revokeToken(refreshToken: String): ApiResult<SuccessResponse> {
        return post("$BASE_URL/api/mobile/revoke", null, mapOf("refreshToken" to refreshToken))
    }

    // MARK: - HTTP Methods

    private suspend inline fun <reified T> get(url: String, token: String?): ApiResult<T> {
        return try {
            val response = httpClient.get(url) {
                token?.let { header("Authorization", "Bearer $it") }
            }
            handleResponse(response)
        } catch (e: Exception) {
            ApiResult.Error(ApiException.NetworkError(e))
        }
    }

    private suspend inline fun <reified T> post(url: String, token: String?, body: Any?): ApiResult<T> {
        return try {
            val response = httpClient.post(url) {
                token?.let { header("Authorization", "Bearer $it") }
                body?.let {
                    contentType(ContentType.Application.Json)
                    setBody(it)
                }
            }
            handleResponse(response)
        } catch (e: Exception) {
            ApiResult.Error(ApiException.NetworkError(e))
        }
    }

    private suspend inline fun <reified T> patch(url: String, token: String?, body: Any?): ApiResult<T> {
        return try {
            val response = httpClient.patch(url) {
                token?.let { header("Authorization", "Bearer $it") }
                body?.let {
                    contentType(ContentType.Application.Json)
                    setBody(it)
                }
            }
            handleResponse(response)
        } catch (e: Exception) {
            ApiResult.Error(ApiException.NetworkError(e))
        }
    }

    private suspend inline fun <reified T> delete(url: String, token: String?, body: Any?): ApiResult<T> {
        return try {
            val response = httpClient.delete(url) {
                token?.let { header("Authorization", "Bearer $it") }
                body?.let {
                    contentType(ContentType.Application.Json)
                    setBody(it)
                }
            }
            handleResponse(response)
        } catch (e: Exception) {
            ApiResult.Error(ApiException.NetworkError(e))
        }
    }

    private suspend inline fun <reified T> handleResponse(response: HttpResponse): ApiResult<T> {
        return when (response.status.value) {
            in 200..299 -> {
                try {
                    ApiResult.Success(response.body())
                } catch (e: Exception) {
                    val bodyText = try { response.bodyAsText() } catch (_: Exception) { "unknown" }
                    Log.e(TAG, "Failed to parse response: ${e.message}, body: $bodyText")
                    ApiResult.Error(ApiException.Unknown(response.status.value, "Failed to parse response"))
                }
            }
            401 -> {
                val errorBody = try {
                    response.body<ApiErrorResponse>()
                } catch (e: Exception) {
                    null
                }
                if (errorBody?.error?.contains("expired") == true) {
                    ApiResult.Error(ApiException.TokenExpired)
                } else {
                    ApiResult.Error(ApiException.Unauthorized)
                }
            }
            404 -> ApiResult.Error(ApiException.NotFound)
            400 -> {
                val message = try {
                    response.body<ApiErrorResponse>().error
                } catch (e: Exception) {
                    "Bad request"
                }
                ApiResult.Error(ApiException.BadRequest(message))
            }
            in 500..599 -> {
                val message = try {
                    response.body<ApiErrorResponse>().error
                } catch (e: Exception) {
                    "Server error"
                }
                ApiResult.Error(ApiException.ServerError(message))
            }
            else -> {
                val message = try {
                    response.body<ApiErrorResponse>().error
                } catch (e: Exception) {
                    "Unknown error"
                }
                ApiResult.Error(ApiException.Unknown(response.status.value, message))
            }
        }
    }

    companion object {
        private const val TAG = "ConvexClient"
        // Production Convex deployment URL
        const val BASE_URL = "https://kindhearted-bloodhound-95.convex.site"
        const val SITE_URL = "https://carrelapp.com"
    }
}
